src/charm.py

```diff
@@ -4,27 +4,263 @@
         self.framework.observe(self.on.mysql_pebble_ready, self._on_mysql_pebble_ready)
+        self.framework.observe(self.on.leader_elected, self._on_leader_elected)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on[PEER].relation_joined, self._on_peer_relation_joined)
+        # Actions events
+        self.framework.observe(
+            self.on.get_generated_passwords_action, self._get_generated_passwords
+        )
+
+    @property
+    def _peers(self):
+        """Retrieve the peer relation (`ops.model.Relation`)."""
+        return self.model.get_relation(PEER)
+
+    @property
+    def _mysql(self):
+        """Returns an instance of the MySQL object from mysqlsh_helpers."""
+        peer_data = self._peers.data[self.app]
+
+        return MySQL(
+            self._get_hostname_by_unit(self.unit.name),
+            peer_data["cluster-name"],
+            peer_data["root-password"],
+            "serverconfig",
+            peer_data["server-config-password"],
+            "clusteradmin",
+            peer_data["cluster-admin-password"],
+            self.unit.get_container("mysql"),
+        )
+
+    @property
+    def _pebble_layer(self) -> Layer:
+        """Return a layer for the pebble service."""
+        return Layer(
+            {
+                "summary": "mysqld layer",
+                "description": "pebble config layer for mysqld",
+                "services": {
+                    self._mysqld_service: {
+                        "override": "replace",
+                        "summary": "mysqld",
+                        "command": "mysqld",
+                        "startup": "enabled",
+                        "user": "mysql",
+                        "group": "mysql",
+                    }
+                },
+            }
+        )
+
+    def _get_hostname_by_unit(self, unit_name: str) -> str:
+        """Create a DNS name for a unit.
+
+        Translate juju unit name to resolvable hostname.
+
+        Args:
+            unit_name: unit name
+        Returns:
+            A string representing the hostname of the unit.
+        """
+        unit_id = unit_name.split("/")[1]
+        return f"{self.app.name}-{unit_id}.{self.app.name}-endpoints"
+
+    # =========================================================================
+    # Charm event handlers
+    # =========================================================================
+
+    def _on_config_changed(self, _) -> None:
+        """Handle the config changed event."""
+        # Only execute on unit leader
+        if not self.unit.is_leader():
+            return
+
+        # Set the cluster name in the peer relation databag if it is not already set
+        peer_data = self._peers.data[self.app]
+
+        if not peer_data.get("cluster-name"):
+            peer_data["cluster-name"] = (
+                self.config.get("cluster-name") or f"cluster-{generate_random_hash()}"
```

Comment:
The config name is different from the name on `config.yaml`.
```suggestion
                self.config.get("cluster_name") or f"cluster-{generate_random_hash()}"
```